Essentials/src/main/java/com/earth2me/essentials/UserData.java

```diff
@@ -441,30 +444,63 @@ public void setJail(final String jail) {
         config.save();
     }
 
-    private List<String> _getMails() {
-        return config.getStringList("mail");
+    private JsonArray _getMails() {
+        return ess.getMail().getMailArray(config.getString("mail"));
     }
 
+    @Deprecated
     public List<String> getMails() {
+        final List<String> list = new ArrayList<>();
+        if (getMailAmount() != 0) {
+            for (JsonElement mail : getMailArray()) {
+                list.add(mail.getAsJsonObject().get("message").getAsString());
+            }
+        }
+        return list;
+    }
+
+    public JsonArray getMailArray() {
         return mails;
     }
 
-    public void setMails(List<String> mails) {
-        if (mails == null) {
+    public int getMailAmount() {
+        return mails == null ? 0 : mails.size();
+    }
+
+    public int getUnreadMailAmount() {
+        if (mails == null || mails.size() == 0) {
+            return 0;
+        }
+
+        int unread = 0;
+        for (JsonElement element : mails) {
+            if (!element.getAsJsonObject().get("read").getAsBoolean()) {
+                unread++;
+            }
+        }
+        return unread;
+    }
+
+    public void setMailArray(JsonArray mailArray) {
+        final JsonObject obj = ess.getMail().getMailObject(mailArray);
+        if (obj == null) {
             config.removeProperty("mail");
-            mails = _getMails();
+            mails = null;
         } else {
-            config.setProperty("mail", mails);
+            config.setProperty("mail", obj.toString());
+            mails = _getMails();
         }
-        this.mails = mails;
         config.save();
     }
 
-    public void addMail(final String mail) {
-        mails.add(mail);
-        setMails(mails);
+    @Deprecated
+    public void setMails(final List<String> mails) {
+        throw new UnsupportedOperationException("UserData#setMails is deprecated and can no longer be used. Please tell the plugin author to update this!");
     }
 
+    @Deprecated
+    abstract void addMail(String mail);
```

Comment:
I'm not sure why this isn't still just implemented in `UserData`?

Reply:
Due to the way the system works, we need an IUser, something UserData isn't. Doesn't cause any side effects so /shrug.